nni/retiarii/operation_def/torch_op_def.py

```diff
@@ -1,46 +1,331 @@
+        if inputs_value[2] is not None:
+            layout_str = f', layout=torch.strided'
+            print('Warning: only support `torch.strided` for now!!!')
+        else:
+            layout_str = ''
+        device_str = f', device=torch.device({inputs[3]})' if inputs_value[3] is not None else ''
+        mem_format_str = f', memory_format={mem_format[inputs_value[5]]}' if inputs_value[5] is not None else ''
+        return f'{output} = torch.{op_name}({inputs[0]}{dtype_str}{layout_str}{device_str}{mem_format_str})'
+
+class AtenRand(PyTorchOperation):
+    _ori_type_name = ['aten::rand', 'aten::randn']
+    def to_forward_code(self, field: str, output: str, inputs: List[str], inputs_value: List[Any] = None) -> str:
+        op_name = self.type.split('::')[-1]
+        dtype_str = f', dtype={scalar_type_to_pytorch_type[inputs_value[1]]}' if inputs_value[1] is not None else ''
+        if inputs_value[2] is not None:
+            layout_str = f', layout=torch.strided'
+            print('Warning: only support `torch.strided` for now!!!')
+        else:
+            layout_str = ''
+        device_str = f', device=torch.device({inputs[3]})' if inputs_value[3] is not None else ''
+        return f'{output} = torch.{op_name}({inputs[0]}{dtype_str}{layout_str}{device_str})'
+
+class AtenFloordiv(PyTorchOperation):
+    _ori_type_name = ['aten::floordiv']
+    def to_forward_code(self, field: str, output: str, inputs: List[str], inputs_value: List[Any] = None) -> str:
+        return f'{output} = {inputs[0]} // {inputs[1]}'
+
+class AtenLen(PyTorchOperation):
+    _ori_type_name = ['aten::len']
+    def to_forward_code(self, field: str, output: str, inputs: List[str], inputs_value: List[Any] = None) -> str:
+        return f'{output} = len({inputs[0]})'
+
+ManuallyChooseDef = {
+    'aten::flatten': [('start_dim', 'int', '0'), ('end_dim', 'int', '-1')],
+    'aten::split': [('split_size', 'int', 'None'), ('dim', 'int', '0')]
+}
+
+TensorOpExceptions = {
+    'aten::sub': lambda output, inputs: f'{output} = {inputs[0]} - {inputs[1]}' # example: x.size(1) - 3
+}
+
+def _get_tensor_ops():
+    def hidden(name):
+        return name.startswith('_') and not name.startswith('__')
+
+    def is_tensor_method(schema):
+        if len(schema.arguments) == 0:
+            return False
+        self = schema.arguments[0]
+        if self.name != 'self':
+            return False
+        if not self.type.isSubtypeOf(torch._C.TensorType.get()):
+            return False
+        return True
+
+    def emit_args(args):
+        # filter out the `out` argument here
+        return [(arg.name, str(arg.type), str(arg.default_value)) for arg in args] #  if arg.name != 'out'
+
+    op_names = []
+    op_args = {}
+    # discover methods
+    for elem in dir(torch.Tensor):
+        if not hidden(elem):
+            schemas = torch._C._jit_get_schemas_for_operator("aten::" + elem)
```

Comment:
This API looks dangerous to me...

Reply:
yes, kind of. this one is copied from this file: https://github.com/pytorch/pytorch/blob/master/torch/jit/supported_ops.py